crates/noirc_evaluator/src/brillig/brillig_ir/registers.rs

```diff
@@ -0,0 +1,55 @@
+use acvm::acir::brillig_vm::RegisterIndex;
+
+use super::ReservedRegisters;
+
+/// Every brillig stack frame/call context has its own view of register space.
+/// This is maintained by copying these registers to the stack during calls and reading them back.
+///
+/// Each has a stack base pointer from which all stack allocations can be offset.
+pub(crate) struct BrilligRegistersContext {
+    /// A free-list of registers that have been deallocated and can be used again.
+    /// TODO(AD): currently, register deallocation is only done with immediate values.
+    /// TODO(AD): See https://github.com/noir-lang/noir/issues/1720
+    deallocated_registers: Vec<RegisterIndex>,
+    /// A usize indicating the next un-used register.
+    next_free_register_index: usize,
+}
+
+impl BrilligRegistersContext {
+    /// Initial register allocation
+    pub(crate) fn new() -> BrilligRegistersContext {
+        BrilligRegistersContext {
+            deallocated_registers: Vec::new(),
+            next_free_register_index: ReservedRegisters::len(),
+        }
+    }
+    /// Lazily iterate over the used registers,
+    /// counting to next_free_register_index while excluding deallocated and reserved registers.
+    /// TODO(AD): unused
+    pub(crate) fn _used_registers_iter(&self) -> impl Iterator<Item = RegisterIndex> + '_ {
```

Comment:
Why does this start with an underscore?

Reply:
Sure, just was hoping to use it soon

Reply:
Point is I think we need to do this and this is how it should be done, and I am trying to get the pieces in